internal/fileutil/fileutil.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package fileutil
+
+import (
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+)
+
+// CreateAndSyncFileAtomically writes the content to the tmpFile, fsyncs the tmpFile,
+// renames the tmpFile to the finalFile. In other words, in the event of a crash, either
+// the final file will not be visible or it will have the full contents.
+// The tmpFile should not be existing. The finalFile, if exists, will be overwritten (default rename behavior)
+func CreateAndSyncFileAtomically(dir, tmpFile, finalFile string, content []byte, perm os.FileMode) error {
+	tempFilePath := filepath.Join(dir, tmpFile)
+	finalFilePath := filepath.Join(dir, finalFile)
+	if err := CreateAndSyncFile(tempFilePath, content, perm); err != nil {
+		return err
+	}
+	if err := os.Rename(tempFilePath, finalFilePath); err != nil {
+		return err
+	}
+	return nil
```

Comment:
Shouldn't we do `syncDir(dir)` here after the rename? -- https://www.slideshare.net/nan1nan1/eat-my-data

Reply:
I left it to the user to call this explicitly. As I do [here](https://github.com/hyperledger/fabric/pull/1457/files#diff-32e6124e2ea1f17d76a537c748cdd5a0R201). 

If we do implicit here then it would be better to do in [this function](https://github.com/hyperledger/fabric/pull/1457/files#diff-3f40cb6523f87e0233140718900e5247R35) as well, in order to maintain a uniform behavior but the latter function may sometime not need it based on the context. So, I decided to leave up to the consumer to explicitly invoke it - So, does the function name convey. Atomicity and durability are two different dimensions. syncDir is more for the durability.

Reply:
Thanks, Manish, for the offline discussion. https://man7.org/linux/man-pages/man2/fsync.2.html clarified some details for me. So both APIs ensures that the data is written to the disk and sync of parent directory is left to the caller.